### PR TITLE
Filesystem based MainDomLock & extract interface for MainDomKey generation

### PR DESCRIPTION
**Pull request by p-m-j**
Fixes for #11935

To test, give the issue a good read and follow the reproduction steps.

Should be able to reproduce the issue with SqlMainDomLock if a discriminator isn't set e.g.

```json
"Umbraco": {
   "CMS": {
       "Global": {
           "MainDomLock": "SqlMainDomLock"
       }
   }
}
```
To confirm fix, add a slot specific MainDomKeyDiscriminator value and try again
```json
"Umbraco": {
   "CMS": {
       "Global": {
           "MainDomLock": "SqlMainDomLock",
           "MainDomKeyDiscriminator": "{{slot name - set this in azure config slot specific}}"
       }
   }
}
```

Please also confirm issue doesn't occur for FileSystemMainDomLock (doesn't require discriminator, should just work OOTB)

```json
"Umbraco": {
   "CMS": {
       "Global": {
           "MainDomLock": "FileSystemMainDomLock"
       }
   }
}
```

Pushed pre-release to myget

```xml

// NuGet.config
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="MyGetUmbracoNightly" value="https://www.myget.org/F/umbraconightly/api/v3/index.json" />
  </packageSources>
</configuration>

// Your csproj
<PackageReference Include="Umbraco.Cms" Version="9.4.0-preview20220222.84372" />
```

I tried both out on azure

### SqlMainDomLock (with discriminator)
image

### FileSystemMainDomLock
image

To speed things up if desired there is a sample project @ https://github.com/p-m-j/what-the-maindom 
Just set configuration as necessary e.g. ServerRole, MainDomLock (mode), MainDomKeyDiscriminator

## Also

Works on linux
image

### CODE REVIEW

**Review by nikolajlauridsen — COMMENTED** [review]
Code looks good to me, only found a couple of constructors I'm guessing can be obsoleted. 

Still need to do some testing though 👍

File: src/Umbraco.Infrastructure/HostedServices/ScheduledPublishing.cs [hunk 1]
```diff
@@ -27,20 +29,13 @@ public class ScheduledPublishing : RecurringHostedServiceBase
...
         /// </summary>
-        /// <param name="runtimeState">Representation of the state of the Umbraco runtime.</param>
-        /// <param name="mainDom">Representation of the main application domain.</param>
-        /// <param name="serverRegistrar">Provider of server registrations to the distributed cache.</param>
-        /// <param name="contentService">Service for handling content operations.</param>
-        /// <param name="umbracoContextFactory">Service for creating and managing Umbraco context.</param>
-        /// <param name="logger">The typed logger.</param>
-        /// <param name="serverMessenger">Service broadcasting cache notifications to registered servers.</param>
-        /// <param name="backofficeSecurityFactory">Creates and manages <see cref="IBackOfficeSecurity"/> instances.</param>
         public ScheduledPublishing(
```
[the remark this answers is not in the document]
Comment (p-m-j) [reply src/Umbraco.Infrastructure/HostedServices/ScheduledPublishing.cs:42]: Resolved 674a51ad

File: src/Umbraco.Infrastructure/Runtime/SqlMainDomLock.cs [hunk 2]
```diff
@@ -51,23 +52,16 @@ public class SqlMainDomLock : IMainDomLock
...
-        {
-            // unique id for our appdomain, this is more unique than the appdomain id which is just an INT counter to its safer
-            _lockId = Guid.NewGuid().ToString();
-            _logger = logger;
-            _globalSettings = globalSettings;
-            _sqlServerSyntax = new SqlServerSyntaxProvider(_globalSettings);
-            _hostingEnvironment = hostingEnvironment;
-            _dbFactory = new UmbracoDatabaseFactory(
-                loggerFactory.CreateLogger<UmbracoDatabaseFactory>(),
+            : this(
```
[the remark this answers is not in the document]
Comment (p-m-j) [reply src/Umbraco.Infrastructure/Runtime/SqlMainDomLock.cs:58]: resolved 674a51ad

**Review by nikolajlauridsen — APPROVED** [review]
Tests out great 👍 

I was able to replicate the issue with slot swaps and confirm that this fixes the issue, I tried using both:

* SqlMainDomLock - with different `Umbraco__CMS__Global__MainDomKeyDiscriminator` environment variables in deploy and prod
* FileSystemMainDomLock  - works OOTB

Both fixes the issue with maindomlock and updates from backoffice is now picked up regardless of slot swapping 👍

**Pull request merged**